File: frontend/src/shared/ipc/viewport.ts
import { invoke } from '@tauri-apps/api/core';

export interface SetViewportArgs {
  zoom: number;
  x: number;
  y: number;
  width: number;
  height: number;
}

export async function setViewport(args: SetViewportArgs): Promise<void> {
  return invoke<void>('set_viewport', {
    zoom: args.zoom,
    x: args.x,
    y: args.y,
    width: args.width,
    height: args.height,
  });
}

File: src-tauri/src/commands/viewport.rs
use std::sync::Arc;
use tauri::State;

use crate::commands::AppState;
use crate::services::ViewportService;
pub use crate::viewport::SetViewportResponse;

/// Set the current viewport state.
#[tauri::command]
pub fn set_viewport(
    zoom: f64,
    x: f64,
    y: f64,
    width: f64,
    height: f64,
    state: State<'_, Arc<AppState>>,
) -> Result<SetViewportResponse, String> {
    ViewportService::new(state.inner().clone())
        .set_viewport(zoom, x, y, width, height)
        .map_err(|e| e.to_string())
}
